### PR TITLE
Cdsk 1163 jwt token integration

### PR DESCRIPTION
**Pull request by gaurav-plivo**
- applied validation for jwt header 
- pass X-Plivo-jwt header if logged by jwt 
- pass X-Plivo-jwt header on invite
- set sub if not present in token
- pass X-Plivo-jwt on invite
- set accesstoken param for test case
- apply unit test cases for jwt

### CODE REVIEW

File: lib/client.ts [hunk 1]
```diff
@@ -691,6 +768,136 @@ export class Client extends EventEmitter {
...
+
+    let { app = undefined, lss = undefined, sub = undefined, nbf = undefined, exp = undefined, per = undefined } = parsedToken;
+    let { incoming_allow = undefined, outgoing_allow = undefined } = per.voice || undefined;
+  
+    // To do : Add the token validations [DONE]
+    if(!lss || !nbf || !exp || !per || incoming_allow === undefined || outgoing_allow === undefined) {
+      return false;
+    }
+    
+    if(typeof app !== "string" || typeof lss !== "string" || typeof sub !== "string" || typeof nbf !== "number" ||
```
Comment (vikash-plivo) [inline lib/client.ts:793]: its not lss. it should be iss. please correct it everywhere

File: lib/managers/account.ts [hunk 2]
```diff
@@ -325,7 +340,14 @@ class Account {
   /**
    * Triggered when the user is logged in.
    */
-  private _onRegistered = (): void => {
+  private _onRegistered = (res: any): void => {
+    // Parse the response to get the JWT expiry in epoch
+    // below is the example to get basic 120 sec expiry from response
+    // To do : This needs to be changed in case of login through access Token method
+    if (this.cs.isAccessToken) {
+      const expiryTimeInEpoch = Number(res['response']['headers']['Contact'][0]['parsed']['_parameters']['expires']);
```
Comment (vikash-plivo) [inline lib/managers/account.ts:348]: setting expiry from contact header? then why do we need extra header from server side?

**Review by vikash-plivo — CHANGES_REQUESTED** [review]

File: lib/client.ts (hunk 1 above)
Comment (gaurav-plivo) [reply lib/client.ts:793]: the "lss" parameter I received from token when I parse it

File: lib/utils/headers.ts [hunk 3]
```diff
@@ -4,14 +4,17 @@
  * @param {String} key - extra header key
  */
 export const checkExtraHeaderKey = function (key: string): boolean {
+  
   const keyUppercase = key.toUpperCase();
-  if (!keyUppercase.startsWith('X-PH')) {
-    return false;
+  
+  if(keyUppercase.startsWith('X-PH') || keyUppercase.startsWith('X-PL')) {
```
Comment (sanyamjain-plivo) [inline lib/utils/headers.ts:10]: add the exact match condition for x-plivo-jwt

File: lib/utils/headers.ts [hunk 4]
```diff
@@ -20,7 +23,7 @@ export const checkExtraHeaderKey = function (key: string): boolean {
  */
 export const checkExtraHeaderVal = function (value: string): boolean {
   // value only contain [A-Z], [a-z], [0-9] and '%', max length = 48
-  const valRegex = /^([a-z0-9A-Z_\-+()%]){1,120}$/; // +-_() added to Customer headers as per close.io request
```
Comment (sanyamjain-plivo) [inline lib/utils/headers.ts:23]: create seperate function for JWT value validation

File: lib/client.ts (hunk 1 above)
Comment (eniyavanm) [reply lib/client.ts:793]: @gaurav-plivo It is mentioned lss it has to be is in line 793

**Pull request merged**